src/track/serato/beatgrid.cpp

```diff
@@ -471,50 +480,61 @@ void SeratoBeatGrid::setBeats(BeatsPointer pBeats,
             if (pBeatsIterator->hasNext()) {
-                const double positionSecs = streamInfo.getSignalInfo().frames2secs(
-                                                    streamInfo.getSignalInfo().samples2frames(
-                                                            currentBeatPositionSamples)) -
+                const double positionSecs =
+                        streamInfo.getSignalInfo().frames2secsFractional(
+                                currentBeatPositionFrames) -
                         timingOffsetSecs;
                 nonTerminalMarkers.append(
                         std::make_shared<SeratoBeatGridNonTerminalMarker>(positionSecs, 0));
             }
             beatsSinceLastMarker = 0;
         }
         beatsSinceLastMarker++;
-        previousDeltaSamples = currentDeltaSamples;
+        previousDeltaFrames = currentDeltaFrames;
     }
 
     // The track has no beats. This isn't possible because the `pBeatsIterator`
     // check above should have caught this case already.
-    VERIFY_OR_DEBUG_ASSERT(currentBeatPositionSamples != -1) {
+    VERIFY_OR_DEBUG_ASSERT(currentBeatPositionFrames != -1) {
         return;
     }
 
+    // If the beatgrid is constant, i.e. if there is only a single non-terminal
+    // beatgrid marker at the start of the beatgrid and a terminal marker at
+    // the end, then Serato discards the non-terminal marker and just saves the
+    // terminal marker instead.
+    if (nonTerminalMarkers.size() == 1) {
+        nonTerminalMarkers.clear();
+    }
+
     // Update the `beatsSinceLastMarker` of the last non-terminal marker we inserted.
-    DEBUG_ASSERT(!nonTerminalMarkers.isEmpty());
-    DEBUG_ASSERT(beatsSinceLastMarker > 0);
-    const auto pNonTerminalMarker = nonTerminalMarkers.at(nonTerminalMarkers.size() - 1);
-    DEBUG_ASSERT(pNonTerminalMarker);
-    // We need to subtract 1 from `beatsSinceLastMarker`, because at the end of
-    // the last iteration the counter is incremented even though we didn't move
-    // a beat forwards.
-    pNonTerminalMarker->setBeatsTillNextMarker(beatsSinceLastMarker - 1);
+    if (!nonTerminalMarkers.isEmpty()) {
+        DEBUG_ASSERT(beatsSinceLastMarker > 0);
+        const auto pNonTerminalMarker = nonTerminalMarkers.at(nonTerminalMarkers.size() - 1);
+        DEBUG_ASSERT(pNonTerminalMarker);
+        // We need to subtract 1 from `beatsSinceLastMarker`, because at the end of
+        // the last iteration the counter is incremented even though we didn't move
+        // a beat forwards.
+        pNonTerminalMarker->setBeatsTillNextMarker(beatsSinceLastMarker - 1);
+    }
 
     // Finally, create the terminal marker.
     const double positionSecs =
-            streamInfo.getSignalInfo().frames2secs(
-                    streamInfo.getSignalInfo().samples2frames(
-                            currentBeatPositionSamples)) -
+            streamInfo.getSignalInfo().frames2secsFractional(
```

Comment:
This should fix the first debug assertion. From https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/MP3.20Decoding